src/lib/core/CHIPConfig.h

```diff
@@ -644,15 +644,21 @@
 #endif // CHIP_CONFIG_UNAUTHENTICATED_CONNECTION_POOL_SIZE
 
 /**
- * @def CHIP_CONFIG_PEER_CONNECTION_POOL_SIZE
+ * @def CHIP_CONFIG_PEER_SECURE_SESSION_POOL_SIZE
+ *
+ * @brief Defines the size of the pool used for tracking the state of
+ * secure sessions. This controls the maximum number of concurrent
+ * established secure sessions across all supported transports.
+ *
+ * This is sized to cover the sum of the following:
+ *  - At least 3 CASE sessions / fabric (Spec Ref: 4.13.2.8)
+ *  - 1 reserved slot for CASEServer as a responder.
+ *  - 1 reserved slot for PASE.
  *
- * @brief Define the size of the pool used for tracking CHIP
- * Peer connections. This defines maximum number of concurrent
- * device connections across all supported transports.
  */
-#ifndef CHIP_CONFIG_PEER_CONNECTION_POOL_SIZE
-#define CHIP_CONFIG_PEER_CONNECTION_POOL_SIZE 16
-#endif // CHIP_CONFIG_PEER_CONNECTION_POOL_SIZE
+#ifndef CHIP_CONFIG_PEER_SECURE_SESSION_POOL_SIZE
+#define CHIP_CONFIG_PEER_SECURE_SESSION_POOL_SIZE (CHIP_CONFIG_MAX_FABRICS * 3 + 2)
```

Comment:
Sessions don't get de-allocated despite all references to it having gone to 0 (i.e no applications or protocols are currently using that session) precisely because we want a session, once established, to remain resident for as long as it can till it needs to be evicted to make space for another. This ensures that even if there is no current use for a session for interactions, that when there is interest in the future, that a previously established session is already available to be used immediately. This is a reflection of the cost it takes to not just discover session mis-match between peers, but the cost of setting up one as well.

Consequently, even if we backed the pool by `ObjectPool`, we still need an upper bound on the number of sessions we can maximally allocate before we start doing eviction on the existing set. The only difference with an `ObjectPool` backed implementation on heap-based platforms is that it the allocated set will start at 0, and slowly creep up to `CHIP_CONFIG_PEER_SECURE_SESSION_POOL_SIZE` before it stays at that size effectively, for-ever...

Reply:
I'll put out a follow-up PR to actually update support to `ObjectPool`, and as part of that, I'll update the comments for this define to reflect the considerations for applications when setting this value.